tls/src/keys.rs

```diff
@@ -0,0 +1,362 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! TLS based connections
+
+use camino::Utf8PathBuf;
+use dice_verifier::PkiPathSignatureVerifier;
+use ed25519_dalek::pkcs8::PrivateKeyInfo;
+use ed25519_dalek::Signer as EdSigner;
+use ed25519_dalek::Verifier;
+use rustls::{
+    client::danger::HandshakeSignatureValid,
+    pki_types::CertificateDer,
+    sign::{CertifiedKey, Signer, SigningKey},
+    SignatureScheme,
+};
+use secrecy::{DebugSecret, ExposeSecret, Secret};
+use sha3::Digest;
+use slog::{error, info};
+use std::io::prelude::*;
+use std::iter;
+
+use crate::ipcc::Ipcc;
+use crate::Error;
+use std::{fs::File, sync::Arc};
+use x509_cert::{
+    der::{self, Decode, Encode, Reader},
+    Certificate,
+};
+use zeroize::{Zeroize, ZeroizeOnDrop};
+
+#[derive(Debug, Clone)]
+pub enum ResolveSetting {
+    // Use certificates gathered over IPCC
+    Ipcc,
+    // Use specified chain/key
+    Local {
+        priv_key: Utf8PathBuf,
+        cert_chain: Utf8PathBuf,
+    },
+}
+
+#[derive(Debug)]
+pub struct CertResolver {
+    pub log: slog::Logger,
+    pub resolve: ResolveSetting,
+}
+
+impl CertResolver {
+    pub fn new(log: slog::Logger, resolve: ResolveSetting) -> Self {
+        CertResolver { log, resolve }
+    }
+
+    fn load_ipcc_key(&self) -> Result<Arc<CertifiedKey>, crate::Error> {
+        let ipcc = Ipcc::new().map_err(crate::Error::RotRequest)?;
+        let cert_chain_bytes = ipcc.rot_get_tq_cert_chain()?;
+        let mut idx = 0;
+        let mut der_vec = vec![];
+        while idx < cert_chain_bytes.len() {
```

Comment:
Can you please add a comment describing what's happening here. It looks like you are separating out each DER cert into it's own wrapped type and then pushing them onto a vec as required by rustls. It would be good for future readers to know this without having to review all the code.

Reply:
Thanks for the details!